File: python/Data Structures and Algorithms in Python Book/recursion/binary_sum.py
# Time Complexity: O(n) as there are 2n-1 function calls
# Space Complexity: O(log n) 
# As we are dividing the recursive call range into half => depth of recursion = 1 + logn(base 2)
def binary_sum(S, start, stop):
    if start >= stop:               # zero element in the slice
        return 0
    elif start == stop -1:          # one element in the slice
        return S[start]
    else:
        mid = (start + stop) // 2
        return binary_sum(S, start, mid) + binary_sum(S, mid, stop) # first half includes sum till mid -1

if __name__ == "__main__":
    s = list(range(1, 11))
    print(binary_sum(s, 0, len(s)))

File: java/recursion/BinarySum.java
package recursion;

public class BinarySum {
    /** Returns the sum of subarray data[low] through data[high] inclusive. */
    // initially invoked as binarySum(data, 0, n−1).
    // Time complexity : O(n)
    public static int binarySum(int[] data, int low, int high) {
        if (low > high) // zero element in subarray
            return 0;
        else if (low == high) // one element in subarray
            return data[low];
        else {
            int mid = (low + high) / 2;
            // depth of the recursion is 1+log2 n.
            return binarySum(data, low, mid) + binarySum(data, mid + 1, high);
        }
    }

    public static void main(String[] args) {
        int[] data = { 1, 2, 3, 4, 5, 6, 7, 8 };
        System.out.println(binarySum(data, 0, data.length - 1));
    }
}
